pkg/k8shandler/service_monitor.go

```diff
@@ -0,0 +1,85 @@
+package k8shandler
+
+import (
+	"fmt"
+
+	monitoringv1 "github.com/coreos/prometheus-operator/pkg/client/monitoring/v1"
+	v1alpha1 "github.com/openshift/elasticsearch-operator/pkg/apis/elasticsearch/v1alpha1"
+	"github.com/operator-framework/operator-sdk/pkg/sdk"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// CreateOrUpdateServiceMonitors ensures the existence of ServiceMonitors for Elasticsearch cluster
+func CreateOrUpdateServiceMonitors(dpl *v1alpha1.Elasticsearch) error {
+	serviceMonitorName := fmt.Sprintf("%s-%s", dpl.Name, "cluster")
+	owner := asOwner(dpl)
+
+	err := createOrUpdateServiceMonitor(serviceMonitorName, dpl.Namespace, dpl.Labels, owner)
+	if err != nil {
+		return fmt.Errorf("Failure creating service %v", err)
+	}
+
+	return nil
+}
+
+func createOrUpdateServiceMonitor(serviceMonitorName, namespace string, labels map[string]string, owner metav1.OwnerReference) error {
+	elasticsearchScMonitor := createServiceMonitor(serviceMonitorName, namespace, labels)
+	addOwnerRefToObject(elasticsearchScMonitor, owner)
+	err := sdk.Create(elasticsearchScMonitor)
+	if err != nil && !errors.IsAlreadyExists(err) {
+		return fmt.Errorf("Failure constructing Elasticsearch ServiceMonitor: %v", err)
+	}
+
+	// TODO: handle update
+
+	return nil
+}
+
+func createServiceMonitor(serviceMonitorName, namespace string, labels map[string]string) *monitoringv1.ServiceMonitor {
+	svcMonitor := serviceMonitor(serviceMonitorName, namespace)
+	svcMonitor.Labels = labels
+	labelSelector := metav1.LabelSelector{
+		MatchLabels: labels,
```

Comment:
this in addition with the endpoints, is the most important thing to make this work, I can't tell from the code right now, but you must make sure these labels are the same as the `Service` pointing at the elasticsearch cluster has

Reply:
I don't think we want to scrape the service but the pods directly.  Does prometheus use the service to determine the endpoints?  Won't the service otherwise round robin the pods not gettings us metrics on all the ES node?

Reply:
Yes Prometheus just uses the Service to determine the endpoints.